tests/foreman/performance/common/perf_concurrent_test.py

```diff
@@ -0,0 +1,372 @@
+
+    def _restore_from_savepoint(self, savepoint):
+        """Restore from savepoint"""
+        self.logger.info('---------------------------------')
+        self.logger.info("Reset db from /home/backup/{}".format(savepoint))
+        self.logger.info('---------------------------------')
+        ssh.command('source /root/reset-db.sh /home/backup/{}'
+                    .format(savepoint))
+
+    def _set_num_iterations(self, total_iterations, current_num_threads):
+        """Set # of iterations each thread will conduct
+
+        @para: total_iterations that a test case would conduct
+        @para: current_num_threads
+
+        @example: 5k for 10 threads; 6k for 6 threads
+        """
+        self.num_iterations = total_iterations / current_num_threads
+
+    def _set_bucket_size(self):
+        """Set size for each bucket"""
+        self.bucket_size = self.num_iterations / self.num_buckets
+
+    def _join_all_threads(self, thread_list):
+        """Wait for all threads to complete
+
+        @para: thread_list
+        """
+        for t in thread_list:
+            t.join()
+
+    def _write_raw_csv_file(self, raw_file_name, time_result_dict,
+                            current_num_threads, header):
+        """ Write raw timing ak/del results to csv file"""
+        self.logger.debug("Timing result is: {}".format(time_result_dict))
+        with open(raw_file_name, 'a') as f:
+            writer = csv.writer(f)
+            writer.writerow([header])
+
+            # for each thread, write its head and data
+            for i in range(current_num_threads):
+                writer.writerow([
+                    'client-{}'.format(i)
+                ])
+                writer.writerow(time_result_dict.get('thread-{}'.format(i)))
+            writer.writerow([])
+
+    def _write_raw_att_csv_file(self, raw_file_name, time_result_dict,
+                                current_num_threads, header):
+        """Write raw timing att results to csv file"""
+        self.logger.debug("Timing result is: {}".format(time_result_dict))
+        with open(raw_file_name, 'a') as f:
+            writer = csv.writer(f)
+            writer.writerow([header])
+
+            # for each thread, write its head and data
+            for i in range(current_num_threads):
+                writer.writerow(['client-{}'.format(i)])
+                for index in range(3):
+                    writer.writerow(time_result_dict.get(
+                                    'thread-{}'.format(i))[index])
+                    writer.writerow([])
+
+    def _write_stat_csv_file(self, stat_file_name, time_result_dict,
+                             current_num_threads, header, is_attach):
+        """Generate statistical result of concurrent ak/del/att"""
+        with open(stat_file_name, 'a') as f:
+            writer = csv.writer(f)
+            writer.writerow([header])
+
+            """1) write stat-per-client-bucketized result of ak/del/att"""
```

Comment:
Better to comment like:

```python
# 1) write stat-per-client-bucketized result of ak/del/att
```

The `"""` notation is used for docstrings and they are allowed only on the first line of a module, or the first line of a method or function.